### PR TITLE
[HOLD] Add OpenAPI spec and validate requests against it

### LINKED ISSUE #1216: add openapi documentation
at the time of filing this ticket, the ReST API is still small, as most non-scheduled/non-automated functionality is handled from the rails console.

once done, update the PR template to mention the openapi doc along with the other documentation to possibly update.

### PR DESCRIPTION
**Pull request by mjgiarlo**
Fixes #1216

## Why was this change made?

Includes:
* Add OpenAPI specification (in YAML format)
* Use Committee gem to validate requests (but not yet responses)
* Validate the OpenAPI specification in CI
* Align ObjectsController documentation with implementation
* Adjust specs now that committee is doing some error handling

This PR could go farther by validating responses in addition to requests, and also by removing some error-checking/validation code paths in the controllers that may now effectively be dead by virtue of the work Committee is doing for us, but I opted to get this PR submitted instead due to its size and complexity.

Note that the [drop in coverage](https://coveralls.io/builds/28888801) is related to :point_up:. I'd like to hear what folks think about this.

## Was the usage documentation (e.g. wiki, README, queue or DB specific README) updated?

yes

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -359,6 +361,16 @@ curl -H 'Authorization: Bearer eyJhbGcxxxxx.eyJzdWIxxxxx.lWMJ66Wxx-xx' http://lo
...
+Build image:
+```
+docker build -t suldlss/preservation_catalog:latest .
+```
+
+Publish:
+```
+docker push suldlss/preservation_catalog:latest
+```
+
```
Comment (ndushay) [inline README.md]: should we switch to circleci to do this automatically when PRs are merged to master?   Or use github actions?   If you think so, make a ticket (not critical path ... out of scope for this work cycle)

File: app/controllers/objects_controller.rb [hunk 2]
```diff
@@ -7,13 +7,13 @@
 #  (Note: methods will eventually be ported from sdr-services-app)
 class ObjectsController < ApplicationController
   # return the PreservedObject model for the druid (supplied with druid: prefix)
```
Comment (ndushay) [inline app/controllers/objects_controller.rb]: *does* it require the druid prefix?  I thought we took it either way?

File: config/application.rb [hunk 3]
```diff
@@ -12,11 +12,40 @@
 # you've limited to :test, :development, or :production.
 Bundler.require(*Rails.groups)
 
+class JSONAPIError < Committee::ValidationError
+  def error_body
+    {
+      errors: [
+        { status: id, detail: message }
```
Comment (ndushay) [inline config/application.rb]: `status: id` ?  really?

File: config/application.rb [hunk 4]
```diff
@@ -12,11 +12,40 @@
...
 
+    # accept_request_filter omits OKComputer
+    accept_proc = proc { |request| request.path.start_with?('/v1') }
+    config.middleware.use Committee::Middleware::RequestValidation, schema_path: 'openapi.yml',
+                                                                    strict: true, error_class: JSONAPIError,
+                                                                    accept_request_filter: accept_proc
+    # TODO: we can uncomment this at a later date to ensure we are passing back
+    #       valid responses. Currently, uncommenting this line causes 24 spec
+    #       failures.
+    #
```
Comment (ndushay) [inline config/application.rb]: make a ticket for this and reference it?

File: docs/index.html [hunk 5]
```diff
@@ -0,0 +1,17 @@
...
+             margin: 0;
+             padding: 0;
+         }
+        </style>
+    </head>
+    <body>
+        <redoc spec-url='https://raw.githubusercontent.com/sul-dlss/preservation_catalog/master/openapi.yml'></redoc>
+        <script src="https://cdn.jsdelivr.net/npm/redoc@2.0.0-alpha.17/bundles/redoc.standalone.js"> </script>
+    </body>
+</html>
```
Comment (ndushay) [inline docs/index.html]: Is this autogen, or required to be provided?  If former, it shouldn't be checked in.  I assume it's the latter. ...

File: openapi.yml [hunk 6]
```diff
@@ -0,0 +1,578 @@
...
+          description: Internal server error
+      requestBody:
+        required: true
+        content:
+          application/x-www-form-urlencoded:
+            schema:
+              type: object
+              properties:
+                druid:
+                  $ref: '#/components/schemas/Druid'
```
Comment (ndushay) [inline openapi.yml]: does this take `bb111bb2222` as well as `druid:bb111bb2222` ?  because prescat does ...

File: openapi.yml [hunk 7]
```diff
@@ -0,0 +1,578 @@
...
+                incoming_version:
+                  type: integer
+                incoming_size:
+                  type: integer
+                storage_location:
+                  type: string
```
Comment (ndushay) [inline openapi.yml]: is there a way to indicate that this must correspond to a real MoabStorageRoot?

File: openapi.yml [hunk 8]
```diff
@@ -0,0 +1,578 @@
...
+      env:
+        default: stage
+  - url: 'https://preservation-catalog-{env}-01.stanford.edu'
+    description: QA service
+    variables:
+      env:
+        default: qa
+tags:
+  - name: catalog
+    description: Add and update MOABs in the catalog
```
Comment (ndushay) [inline openapi.yml]: Moab is not an acronym;  it should be Moab everywhere, if we care.

File: openapi.yml [hunk 9]
```diff
@@ -0,0 +1,578 @@
...
+                    - 'false'
+                    - 'False'
+                    - 'FALSE'
+              required:
+                - druid
+                - incoming_version
+                - incoming_size
+                - storage_location
+  /v1/catalog/{druid}:
+    put:
```
Comment (ndushay) [inline openapi.yml]: based on catalog_controller.rb#L28-29, I'm thinking we don't allow PUT?  Or perhaps we don't *mean* to allow it?

File: openapi.yml [hunk 10]
```diff
@@ -0,0 +1,578 @@
+openapi: 3.0.0
```
Comment (ndushay) [inline openapi.yml]: wondering if there's any way to indicate with openapi that an access token is required ...

File: openapi.yml [hunk 11]
```diff
@@ -0,0 +1,578 @@
...
+          in: query
+          required: false
+          example: '9'
+          schema:
+            type: string
+  /v1/objects/{id}/content_diff:
+    post:
+      tags:
+        - objects
+      summary: Return a content metadata diff
```
Comment (ndushay) [inline openapi.yml]: a Moab::FileInventoryDifference for contentMetadata?   It is specific to the Moab ...

File: spec/requests/objects_controller_checksums_spec.rb [hunk 12]
```diff
@@ -201,15 +201,17 @@
         it 'body has additional information from the exception if available' do
           post checksums_objects_url, params: { druids: [], format: :json }, headers: valid_auth_header
           expect(response).to have_http_status(:bad_request)
-          expect(response.body).to eq '400 Bad Request: Identifier has invalid suri syntax:  nil or empty'
+          error_response = JSON.parse(response.body)['errors'].first
+          expect(error_response['detail']).to include('does not match value: , example: druid:bc123df4567')
```
Comment (ndushay) [inline spec/requests/objects_controller_checksums_spec.rb]: we no longer return the error codes?  I think we put them in there so the consumer, via preservation-client, could see them?  Or am I garbling things in my mind?   Are the errors on the consumer side too repetitious as it is???

File: spec/requests/objects_controller_content_diff_spec.rb [hunk 13]
```diff
@@ -64,8 +64,9 @@
           params = { content_metadata: content_md, subset: 'unrecognized' }
           post content_diff_object_url(id: prefixed_druid), params: params, headers: valid_auth_header
           expect(response).to have_http_status(:bad_request)
-          m = "400 Bad Request: subset arg must be 'all', 'shelve', 'preserve', or 'publish' (MoabStorageService.content_diff for druid bj102hs9687)"
-          expect(response.body).to eq m
+          error_response = JSON.parse(response.body)['errors'].first
+          expect(error_response['status']).to eq('bad_request')
+          expect(error_response['detail']).to include('unrecognized isn\'t include enum') # sic
```
Comment (ndushay) [inline spec/requests/objects_controller_content_diff_spec.rb]: ew - yucky error message.  Committee gem giveth and taketh away.

File: config/application.rb [hunk 14]
```diff
@@ -12,11 +12,40 @@
...
+
+  def render
+    [
+      status,
+      { 'Content-Type' => 'application/vnd.api+json' },
+      [JSON.generate(error_body)]
+    ]
+  end
+end
+
```
Comment (ndushay) [inline config/application.rb]: If this changes all the error messages, do we have dead code anywhere (where the code supplied the error messages)

File: spec/requests/objects_controller_file_spec.rb [hunk 15]
```diff
@@ -96,8 +98,9 @@
     context 'when druid invalid' do
       it 'returns 404 response code with "Identifier has invalid suri syntax"' do
```
Comment (ndushay) [inline spec/requests/objects_controller_file_spec.rb]: description no longer true?

File: spec/requests/objects_controller_file_spec.rb [hunk 16]
```diff
@@ -79,8 +80,9 @@
       context 'when id param is empty' do
         it "returns 404 with 'Couldn't find PreservedObject'" do
```
Comment (ndushay) [inline spec/requests/objects_controller_file_spec.rb]: description no longer true?

File: README.md (hunk 1 above)
Comment (mjgiarlo) [reply README.md]: Probably. I'll create a ticket.

Comment (mjgiarlo) [reply README.md]: Boop: https://github.com/sul-dlss/preservation_catalog/issues/1406

File: config/application.rb (hunk 3 above)
Comment (mjgiarlo) [reply config/application.rb]: @ndushay Really: https://github.com/interagent/committee#validation-errors

File: config/application.rb (hunk 14 above)
Comment (mjgiarlo) [reply config/application.rb]: Yes! From the PR desc:
> This PR could go farther by ... removing some error-checking/validation code paths in the controllers that may now effectively be dead by virtue of the work Committee is doing for us

I'd be happy to include that in this PR! I'll push that up in a bit.

File: config/application.rb (hunk 4 above)
Comment (mjgiarlo) [reply config/application.rb]: https://github.com/sul-dlss/preservation_catalog/issues/1407

Comment (mjgiarlo) [reply config/application.rb]: Done.

File: docs/index.html (hunk 5 above)
Comment (mjgiarlo) [reply docs/index.html]: Not auto-generated and is required for API docs like https://sul-dlss.github.io/sdr-api/

File: spec/requests/objects_controller_checksums_spec.rb (hunk 12 above)
Comment (mjgiarlo) [reply spec/requests/objects_controller_checksums_spec.rb]: No, because this is a committee error code. The client should still get the status; it just won't be duplicated in the body/message.

File: spec/requests/objects_controller_content_diff_spec.rb (hunk 13 above)
Comment (mjgiarlo) [reply spec/requests/objects_controller_content_diff_spec.rb]: yeah :unamused:

File: spec/requests/objects_controller_file_spec.rb (hunk 16 above)
Comment (mjgiarlo) [reply spec/requests/objects_controller_file_spec.rb]: fixed! good :eye:

File: spec/requests/objects_controller_file_spec.rb (hunk 15 above)
Comment (mjgiarlo) [reply spec/requests/objects_controller_file_spec.rb]: fixed

File: app/controllers/objects_controller.rb (hunk 2 above)
Comment (mjgiarlo) [reply app/controllers/objects_controller.rb]: Addressed here: https://github.com/sul-dlss/preservation_catalog/pull/1408 Resolving.

File: openapi.yml (hunk 10 above)
Comment (mjgiarlo) [reply openapi.yml]: @ndushay There is! It's at the bottom of the file: https://github.com/sul-dlss/preservation_catalog/pull/1389/files#diff-dc51a340f5f12b9af5a5462818e6eca1R572-R578 Resolving.

File: openapi.yml (hunk 6 above)
Comment (mjgiarlo) [reply openapi.yml]: Yes. I made the `druid:` prefix optional in this schema's regex: https://github.com/sul-dlss/preservation_catalog/pull/1389/files#diff-dc51a340f5f12b9af5a5462818e6eca1R424

Resolving.

File: openapi.yml (hunk 7 above)
Comment (mjgiarlo) [reply openapi.yml]: @ndushay will do.

File: openapi.yml (hunk 8 above)
Comment (mjgiarlo) [reply openapi.yml]: Fixed!

File: openapi.yml (hunk 9 above)
Comment (mjgiarlo) [reply openapi.yml]: I don't think that's the case. Will update that comment.

File: openapi.yml (hunk 11 above)
Comment (mjgiarlo) [reply openapi.yml]: Change made.

**Comment by jmartin-sul** [comment]
>  I would like this to be tested in stage before a merge. Ideally with some bad requests!

agreed!

File: openapi.yml [hunk 17]
```diff
@@ -0,0 +1,578 @@
...
+          - filename
+          - md5
+          - sha1
+          - sha256
+          - filesize
+    PreservedObject:
+      description: A preserved object instance
+      type: object
+      properties:
+        id:
```
[the remark this answers is not in the document]
Comment (jmartin-sul) [reply openapi.yml]: https://github.com/sul-dlss/preservation_catalog/issues/1412

File: config/application.rb [hunk 18]
```diff
@@ -12,11 +12,40 @@
...
 module PreservationCatalog
   class Application < Rails::Application
     # Initialize configuration defaults for originally generated Rails version.
     config.load_defaults 5.1
 
+    # accept_request_filter omits OKComputer
```
[the remark this answers is not in the document]
Comment (mjgiarlo) [reply config/application.rb]: Good :eye:. Got it.

File: docs/index.html [hunk 19]
```diff
@@ -0,0 +1,17 @@
...
+        <style>
+         body {
+             margin: 0;
+             padding: 0;
+         }
+        </style>
+    </head>
+    <body>
+        <redoc spec-url='https://raw.githubusercontent.com/sul-dlss/preservation_catalog/master/openapi.yml'></redoc>
+        <script src="https://cdn.jsdelivr.net/npm/redoc@2.0.0-alpha.17/bundles/redoc.standalone.js"> </script>
```
[the remark this answers is not in the document]
Comment (mjgiarlo) [reply docs/index.html]: Not that I'm aware of. FWIW, this is the same approach we're using for dor-services-app, technical-metadata-service, sdr-api, etc.

File: openapi.yml (hunk 17 above)
[the remark this answers is not in the document]
Comment (mjgiarlo) [reply openapi.yml]: OK, @jmartin-sul. I added a note to that ticket about openapi as well. Thanks!

File: openapi.yml [hunk 20]
```diff
@@ -0,0 +1,578 @@
...
+          example: 13
+        created_at:
+          type: string
+          format: date-time
+          example: '2020-02-20T06:34:16.148Z'
+        updated_at:
+          type: string
+          format: date-time
+          example: '2020-02-21T16:36:28.075Z'
+        preservation_policy_id:
```
[the remark this answers is not in the document]
Comment (mjgiarlo) [reply openapi.yml]: @jmartin-sul mind adding to the other ticket you created (#1412) or creating a companion ticket?

File: openapi.yml [hunk 21]
```diff
@@ -0,0 +1,543 @@
...
+          type: object
+          properties:
+            druid:
+              $ref: '#/components/schemas/Druid'
+            checksums:
+              $ref: '#/components/schemas/ChecksumsForObject'
+    ChecksumsForObject:
+      description: A representation of all files and checksums for a single object
+      type: array
+      minItems: 1
```
[the remark this answers is not in the document]
Comment (mjgiarlo) [reply openapi.yml]: I'll omit it.

File: openapi.yml [hunk 22]
```diff
@@ -0,0 +1,543 @@
...
+            minItems: 1
+            items:
+              $ref: '#/components/schemas/Druid'
+        - name: format
+          in: query
+          required: true
+          example: 'json'
+          schema:
+            type: string
+        - name: return_bare_druids
```
[the remark this answers is not in the document]
Comment (mjgiarlo) [reply openapi.yml]: :beers:

File: openapi.yml [hunk 23]
```diff
@@ -0,0 +1,543 @@
...
+        - report_datetime
+    ChecksumsForObjects:
+      description: A representation of all files and checksums for a list of objects
+      type: array
+      minItems: 1
+      items:
+        type: object
+        additionalProperties:
+          type: object
+          properties:
```
[the remark this answers is not in the document]
Comment (mjgiarlo) [reply openapi.yml]: We are testing this in `spec/requests/objects_controller_checksums_spec.rb`, so I believe this is the right incantation (with druids as keys and the checksums array as the values).

`additionalProperties` is what you use in openapi to handle validation of dicts/maps/hashes that do not have predictably named keys, AFAICT.

File: openapi.yml [hunk 24]
```diff
@@ -0,0 +1,543 @@
...
+              type: string
+            added:
+              type: string
+            deleted:
+              type: string
+            subset:
+              type: array
+              items:
+                type: object
+                properties:
```
[the remark this answers is not in the document]
Comment (mjgiarlo) [reply openapi.yml]: added! thanks for finding that. the moab-versioning stuff is a little harder to follow.

File: openapi.yml [hunk 25]
```diff
@@ -0,0 +1,543 @@
...
+          type: string
+        checksums_validated:
+          type: string
+          enum:
+            - 'true'
+            - 'True'
+            - 'TRUE'
+            - 'nil'
+            - '1'
+            - 'on'
```
[the remark this answers is not in the document]
Comment (mjgiarlo) [reply openapi.yml]: These cases are from https://github.com/sul-dlss/preservation_catalog/blob/master/spec/controllers/catalog_controller_spec.rb#L297-L308 so, yeah, I'm inclined not to tweak the openapi since these are cases we're testing for.

**Comment by mjgiarlo** [comment]
@jmartin-sul Can you take another look? I believe I've addressed or responded to all your points. Thanks!

**Comment by mjgiarlo** [comment]
I'm planning to pair w/ @ndushay tomorrow to test this in stage, btw, assuming we can find some time to dig in.

File: app/controllers/objects_controller.rb [hunk 26]
```diff
@@ -70,8 +63,6 @@ def checksums
       end
       format.any { render status: :not_acceptable, plain: 'Format not acceptable' }
     end
-  rescue Moab::InvalidSuriSyntaxError => e
-    render(plain: "400 Bad Request: #{e}", status: :bad_request)
```
Comment (ndushay) [inline app/controllers/objects_controller.rb]: this is comical b/c we went to some trouble to get this to come from Moab. hahahaha

File: docs/index.html (hunk 19 above)
[the remark this answers is not in the document]
Comment (jmartin-sul) [reply docs/index.html]: no worries, and sorta expected that to be the answer, but thought i'd ask anyway 🙂

File: openapi.yml (hunk 23 above)
[the remark this answers is not in the document]
Comment (jmartin-sul) [reply openapi.yml]: > We are testing this in spec/requests/objects_controller_checksums_spec.rb, so I believe this is the right incantation (with druids as keys and the checksums array as the values).

good point!  not sure i fully grok how to use `additionalProperties` based on the little reading i just did, but i suspect i was indeed mis-understanding how this should be declared, and i trust the tests are exercising this after looking at them.

> _or_ i could be mis-understanding how the yml declares this

left myself an out so my comment could still be right overall 😆

File: openapi.yml (hunk 24 above)
[the remark this answers is not in the document]
Comment (jmartin-sul) [reply openapi.yml]: > the moab-versioning stuff is a little harder to follow

yup!

File: openapi.yml (hunk 25 above)
[the remark this answers is not in the document]
Comment (jmartin-sul) [reply openapi.yml]: ah!  ok.  that totally makes sense.  i hadn't realized we were testing for those odd cases, but i would agree this is a totally reasonable solution.  i think i'd slightly rather that we make the spec more restrictive and then have our tests see `'nil', '1', 'on'` as invalid input instead of falsey options, but that strikes me as a backlog ticket.  don't have strong feelings on that, and certainly wouldn't hold up this PR over that.
